File: ios/Muthal/Models/Summary.swift
import Foundation

/// Ports web lib/summary.ts / android SummaryHelper.kt — keep in sync (SPEC §5).
struct Summary {
    let totalIncome: Double
    let totalExpense: Double
    let balance: Double
    let monthIncome: Double
    let monthExpense: Double
    let monthBalance: Double
}

enum SummaryCalc {
    struct Item {
        let amount: Double
        let type: String
        let date: Date
    }

    /// Month membership decided by the UTC calendar month+year of `now`.
    static func summarize(_ entries: [Item], now: Date) -> Summary {
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let ny = cal.component(.year, from: now)
        let nm = cal.component(.month, from: now)

        var ti = 0.0, te = 0.0, mi = 0.0, me = 0.0
        for e in entries {
            let inMonth = cal.component(.year, from: e.date) == ny
                && cal.component(.month, from: e.date) == nm
            if e.type == "income" {
                ti += e.amount; if inMonth { mi += e.amount }
            } else if e.type == "expense" {
                te += e.amount; if inMonth { me += e.amount }
            }
        }
        return Summary(
            totalIncome: ti, totalExpense: te, balance: ti - te,
            monthIncome: mi, monthExpense: me, monthBalance: mi - me
        )
    }

    /// Opening/closing balance + in-range totals for [start, endInclusive] (SPEC §7).
    /// Entries strictly before the window feed only the opening balance; entries
    /// strictly after the window are excluded entirely.
    static func periodSummarize(_ entries: [Item], start: Date, endInclusive: Date) -> PeriodSummary {
        var opening = 0.0, periodIncome = 0.0, periodExpense = 0.0, count = 0
        for e in entries {
            if e.date < start {
                if e.type == "income" { opening += e.amount }
                else if e.type == "expense" { opening -= e.amount }
            } else if e.date >= start && e.date <= endInclusive {
                count += 1
                if e.type == "income" { periodIncome += e.amount }
                else if e.type == "expense" { periodExpense += e.amount }
            }
        }
        return PeriodSummary(
            openingBalance: opening, periodIncome: periodIncome, periodExpense: periodExpense,
            closingBalance: opening + periodIncome - periodExpense, entryCount: count
        )
    }
}

struct PeriodSummary {
    let openingBalance: Double
    let periodIncome: Double
    let periodExpense: Double
    let closingBalance: Double
    let entryCount: Int
}

File: web/lib/summary.ts
/** Ports android SummaryHelper.kt / ios Summary.swift — keep in sync (SPEC §5). */

export interface Summary {
  totalIncome: number;
  totalExpense: number;
  balance: number;
  monthIncome: number;
  monthExpense: number;
  monthBalance: number;
}

interface SummaryInput {
  amount: number;
  type: string;
  date: number; // epoch ms
}

/**
 * Reduce a list of entries to the hero totals. Month membership is decided by
 * the **UTC** calendar month+year of `now`, so all platforms agree regardless
 * of device timezone (SPEC §5).
 */
export function summarize(entries: SummaryInput[], now: number): Summary {
  const ref = new Date(now);
  const y = ref.getUTCFullYear();
  const m = ref.getUTCMonth();

  let totalIncome = 0;
  let totalExpense = 0;
  let monthIncome = 0;
  let monthExpense = 0;

  for (const e of entries) {
    const d = new Date(e.date);
    const inMonth = d.getUTCFullYear() === y && d.getUTCMonth() === m;
    if (e.type === "income") {
      totalIncome += e.amount;
      if (inMonth) monthIncome += e.amount;
    } else if (e.type === "expense") {
      totalExpense += e.amount;
      if (inMonth) monthExpense += e.amount;
    }
  }

  return {
    totalIncome,
    totalExpense,
    balance: totalIncome - totalExpense,
    monthIncome,
    monthExpense,
    monthBalance: monthIncome - monthExpense,
  };
}

export interface PeriodSummary {
  openingBalance: number;
  periodIncome: number;
  periodExpense: number;
  closingBalance: number;
  entryCount: number;
}

/**
 * Opening/closing balance + in-range totals for [start, endInclusive] (SPEC §7).
 * Entries strictly before the window feed only the opening balance; entries
 * strictly after the window are excluded entirely.
 */
export function periodSummarize(
  entries: SummaryInput[],
  start: number,
  endInclusive: number,
): PeriodSummary {
  let openingBalance = 0;
  let periodIncome = 0;
  let periodExpense = 0;
  let entryCount = 0;

  for (const e of entries) {
    if (e.date < start) {
      if (e.type === "income") openingBalance += e.amount;
      else if (e.type === "expense") openingBalance -= e.amount;
    } else if (e.date >= start && e.date <= endInclusive) {
      entryCount++;
      if (e.type === "income") periodIncome += e.amount;
      else if (e.type === "expense") periodExpense += e.amount;
    }
  }

  return {
    openingBalance,
    periodIncome,
    periodExpense,
    closingBalance: openingBalance + periodIncome - periodExpense,
    entryCount,
  };
}
